ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/AuthHandlers.java

```diff
@@ -52,51 +57,73 @@ static void setupAuthHandlers(final Server server, final Router router,
 
     systemAuthenticationHandler.ifPresent(handler -> router.route().handler(handler));
 
-    if (jaasAuthHandler.isPresent() || authenticationPlugin.isPresent()) {
-      router.route().handler(AuthHandlers::pauseHandler);
+    if (jaas.isPresent() || authenticationPlugin.isPresent()) {
+      router.route().handler(rc -> selectHandler(rc, jaas.isPresent(), pluginHandler.isPresent()));
 
-      router.route().handler(rc -> wrappedAuthHandler(rc, jaasAuthHandler, pluginHandler));
+      // set up using JAAS
+      jaas.ifPresent(h -> router.route().handler(wrappedHandler(h, Provider.JAAS)));
+      router.route().handler(wrappedHandler(
+          new RoleBasedAuthZHandler(
+              server.getConfig().getList(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG)),
+          Provider.JAAS));
 
+      // set up using PLUGIN
+      pluginHandler.ifPresent(h -> router.route().handler(wrappedHandler(h, Provider.PLUGIN)));
+
+      router.route().handler(AuthHandlers::pauseHandler);
       // For authorization use auth provider configured via security extension (if any)
       securityExtension.getAuthorizationProvider()
           .ifPresent(ksqlAuthorizationProvider -> router.route()
               .handler(new KsqlAuthorizationProviderHandler(server.getWorkerExecutor(),
                   ksqlAuthorizationProvider)));
-
       router.route().handler(AuthHandlers::resumeHandler);
     }
   }
 
-  private static void wrappedAuthHandler(final RoutingContext routingContext,
-      final Optional<AuthHandler> jaasAuthHandler,
-      final Optional<Handler<RoutingContext>> pluginHandler) {
-    if (SystemAuthenticationHandler.isAuthenticatedAsSystemUser(routingContext)) {
-      routingContext.next();
-      return;
-    }
-    // Fall through to authing with Jaas
-    if (jaasAuthHandler.isPresent()) {
-      // If we have a Jaas handler configured and we have Basic credentials then we should auth
-      // with that
-      final String authHeader = routingContext.request().getHeader("Authorization");
-      if (authHeader != null && authHeader.toLowerCase().startsWith("basic ")) {
-        jaasAuthHandler.get().handle(routingContext);
-        return;
+  private static Handler<RoutingContext> wrappedHandler(
+      final Handler<RoutingContext> handler,
+      final Provider provider
+  ) {
+    return rc -> {
+      if (rc.data().get(PROVIDER_KEY) != provider) {
+        rc.next();
```

Comment:
nit: can we add a code comment about why this is OK? Took me a number of read-throughs to understand how `selectHandler()` ensures that a valid handler will always be selected, so we'll never end up in a situation where all the handlers are skipped and no auth is performed.

Reply:
👍